src/npcmove.cpp

```diff
@@ -313,15 +313,36 @@ void npc::execute_action( npc_action action )
         break;
 
     case npc_reload: {
-        moves -= weapon.reload_time(*this);
-        if( ! weapon.reload( *this, weapon.pick_reload_ammo( *this, false ) ) ) {
-            debugmsg("NPC reload failed.");
+        // fetch potential ammo from npc possessions only and exclude empty magazines
+        auto ammo = find_ammo( weapon, false, -1 );
+        ammo.erase( std::remove_if( ammo.begin(), ammo.end(), [this]( const item_location& e ) {
+            return e && weapon.can_reload( e->typeId() );
```

Comment:
The `e &&` indicates the `item_location` (being an element of the `ammo` vector) can be invalid. However, in the call to `sort` below, you access the item_locations that are to be compared without that check.

Either `ammo` can never contain invalid locations, in which case the check here is not needed (and implies a wrong assumption to the reader).

Or `ammo` can contain invalid locations, in which case the check is needed in the lambda for `sort` as well.